### PR TITLE
Async ImageLoader

### PR DESCRIPTION
**Pull request by Yancey1989**
This PR can improve the `ImageLoader` performance.

1. only scanning the loader on Mac OS:  `85.798495 samples /sec` -> `115.098278 samples /sec` (+35%)
1. the ResNet training on Nvidia P100 GPU: `35 samples/sec ` -> `80 samples/sec` (+128%)

### CODE REVIEW

File: vision/datasets/imageloader.go [hunk 1]
```diff
@@ -4,99 +4,117 @@ import (
...
 		label := p.vocab[classStr]
-		p.labels = append(p.labels, label)
 
 		m, _, err := image.Decode(p.r)
 		if err != nil {
-			p.err = err
+			p.err <- err
 			break
 		}
 		input := p.trans.Run(m)
```
Comment (QiJune) [inline vision/datasets/imageloader.go:80]: There are tensor operations in transforms. So, when the tensors are freed?

Comment (Yancey1989) [reply vision/datasets/imageloader.go]: I added a function `splitComposeByToTensor` to splite the composeTransoform into `trans1` and `trans2` by  `ToTensorTransform`. `trans1` running in the goroutine, and `trans2` running at retrieving the minibatch. So we can free the Tensors at each iteration.

File: vision/datasets/imageloader.go [hunk 2]
```diff
@@ -10,74 +10,100 @@ import (
...
-		vocab:  vocab,
-		err:    nil,
-		trans:  trans,
-		mbSize: mbSize,
-	}, nil
+	trans1, trans2 := splitComposeByToTensor(trans)
+	m := &ImageLoader{
+		r:       r,
+		vocab:   vocab,
+		samples: make(chan sample, mbSize*4),
```
Comment (shendiaomo) [inline vision/datasets/imageloader.go:41]: Do we need a buffered channel?

Comment (Yancey1989) [reply vision/datasets/imageloader.go:41]: I think so, the buffered channel can get better performance,  `dataloade`r can reading data at training.

**Comment by Yancey1989** [comment]
Using a buffered channel in `Dataloader.samples` would get a bug as the following logs, so this PR only use the unbuffered channel, we should fix this bug and use the buffered channel to get better performance.

``` text
2020/09/08 15:14:13 No CUDA found; CPU only
2020/09/08 15:14:13 building label vocabulary done.
panic: Target 17781164609562 is out of bounds.
Exception raised from nll_loss_out_frame at ../aten/src/ATen/native/LossNLL.cpp:112 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >) + 169 (0x43fb639 in libc10.dylib)
frame #1: at::native::(anonymous namespace)::nll_loss_forward_out_cpu_template(at::Tensor&, at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)::$_0::operator()() const + 7218 (0x9500c12 in libtorch_cpu.dylib)
frame #2: at::native::nll_loss_forward_out_cpu(at::Tensor&, at::Tensor&, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 309 (0x94fb3e5 in libtorch_cpu.dylib)
frame #3: at::native::nll_loss_forward_cpu(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 167 (0x94fba27 in libtorch_cpu.dylib)
frame #4: at::CPUType::nll_loss_forward(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 14 (0x98fe1fe in libtorch_cpu.dylib)
frame #5: c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<std::__1::tuple<at::Tensor, at::Tensor> (*)(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long), std::__1::tuple<at::Tensor, at::Tensor>, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long> >, std::__1::tuple<at::Tensor, at::Tensor> (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)>::call(c10::OperatorKernel*, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 34 (0x995b352 in libtorch_cpu.dylib)
frame #6: std::__1::tuple<at::Tensor, at::Tensor> c10::Dispatcher::call<std::__1::tuple<at::Tensor, at::Tensor>, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long>(c10::TypedOperatorHandle<std::__1::tuple<at::Tensor, at::Tensor> (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)> const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) const + 297 (0x9a78709 in libtorch_cpu.dylib)
frame #7: at::nll_loss_forward(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 167 (0x99b2497 in libtorch_cpu.dylib)
frame #8: torch::autograd::VariableType::nll_loss_forward(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 428 (0xba5abfc in libtorch_cpu.dylib)
frame #9: c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<std::__1::tuple<at::Tensor, at::Tensor> (*)(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long), std::__1::tuple<at::Tensor, at::Tensor>, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long> >, std::__1::tuple<at::Tensor, at::Tensor> (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)>::call(c10::OperatorKernel*, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 34 (0x995b352 in libtorch_cpu.dylib)
frame #10: std::__1::tuple<at::Tensor, at::Tensor> c10::Dispatcher::call<std::__1::tuple<at::Tensor, at::Tensor>, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long>(c10::TypedOperatorHandle<std::__1::tuple<at::Tensor, at::Tensor> (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)> const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) const + 297 (0x9a78709 in libtorch_cpu.dylib)
frame #11: at::nll_loss_forward(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 167 (0x99b2497 in libtorch_cpu.dylib)
frame #12: at::native::nll_loss(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 23 (0x94fef07 in libtorch_cpu.dylib)
frame #13: at::TypeDefault::nll_loss(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 143 (0x9af6b0f in libtorch_cpu.dylib)
frame #14: torch::autograd::VariableType::nll_loss(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 14 (0xb987b2e in libtorch_cpu.dylib)
frame #15: c10::impl::wrap_kernel_functor_unboxed_<c10::impl::detail::WrapFunctionIntoRuntimeFunctor_<at::Tensor (*)(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long), at::Tensor, c10::guts::typelist::typelist<at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long> >, at::Tensor (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)>::call(c10::OperatorKernel*, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 34 (0x92b8bf2 in libtorch_cpu.dylib)
frame #16: at::Tensor c10::Dispatcher::call<at::Tensor, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long>(c10::TypedOperatorHandle<at::Tensor (at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long)> const&, at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) const + 297 (0x9a778b9 in libtorch_cpu.dylib)
frame #17: at::nll_loss(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, long long) + 167 (0x99b22e7 in libtorch_cpu.dylib)
frame #18: torch::nn::functional::detail::nll_loss(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, c10::variant<torch::enumtype::kNone, torch::enumtype::kMean, torch::enumtype::kSum>) + 218 (0x4383a5a in libcgotorch.so)
frame #19: torch::nn::functional::detail::cross_entropy(at::Tensor const&, at::Tensor const&, at::Tensor const&, long long, c10::variant<torch::enumtype::kNone, torch::enumtype::kMean, torch::enumtype::kSum>) + 131 (0x4383793 in libcgotorch.so)
frame #20: CrossEntropy + 460 (0x437fc4c in libcgotorch.so)
frame #21: _cgo_13f6f2af77f0_Cfunc_CrossEntropy + 49 (0x416cfd1 in resnet)
frame #22: runtime.asmcgocall + 112 (0x4068930 in resnet)


goroutine 1 [running]:
github.com/wangkuiyi/gotorch.MustNil(0x3e09b800)
	/Users/yancey/go/src/github.com/wangkuiyi/gotorch/tensor.go:26 +0x97
github.com/wangkuiyi/gotorch/nn/functional.CrossEntropy(0xc00007e740, 0xc0003520b0, 0x0, 0xffffffffffffff9c, 0x41bb621, 0x4, 0x410fdea)
	/Users/yancey/go/src/github.com/wangkuiyi/gotorch/nn/functional/functional.go:160 +0x9e
main.trainOneMinibatch(0xc0003520a8, 0xc0003520b0, 0xc0000de090, 0xc0000e8a00, 0x1, 0xc0003520b0)
	/Users/yancey/go/src/github.com/wangkuiyi/gotorch/example/resnet/resnet.go:92 +0x85
main.train(0x7ffeefbffa01, 0x3b, 0x7ffeefbffa43, 0x2c, 0x41bf0e8, 0x17, 0x5)
	/Users/yancey/go/src/github.com/wangkuiyi/gotorch/example/resnet/resnet.go:153 +0x33a
main.main()
	/Users/yancey/go/src/github.com/wangkuiyi/gotorch/example/resnet/resnet.go:198 +0x2bc
exit status 2
```

File: vision/datasets/imageloader_test.go [hunk 3]
```diff
@@ -47,3 +86,71 @@ func TestImageTgzLoader(t *testing.T) {
 	_, e = BuildLabelVocabularyFromTgz("no file")
 	a.Error(e)
 }
+
+func TestImageTgzLoaderHeavy(t *testing.T) {
+	if os.Getenv("GOTORCH_TEST_IMAGE_TGZ_PATH") == "" {
```
Comment (wangkuiyi) [inline vision/datasets/imageloader_test.go:91]: Add a comment here on how can we generate such a big tarball?  For example, download it from somewhere and/or shuffle it using `tools/tarball_{split|merge}`?

File: vision/datasets/imageloader.go [hunk 4]
```diff
@@ -10,75 +10,104 @@ import (
...
-		trans:  trans,
-		mbSize: mbSize,
-	}, nil
+	trans1, trans2 := splitComposeByToTensor(trans)
+	m := &ImageLoader{
+		r:     r,
+		vocab: vocab,
+		// TODO(yancey1989): using a buffered channel to get better performance, but
+		// a bug should be fixed before. c.f. https://github.com/wangkuiyi/gotorch/pull/304#issuecomment-688670165
+		samples: make(chan sample),
```
Comment (QiJune) [inline vision/datasets/imageloader.go:44]: Since FromBlob is using deep copy now, we could have a test on the buffered channel.

Comment (Yancey1989) [reply vision/datasets/imageloader.go]: done

File: vision/datasets/imageloader_test.go (hunk 3 above)
Comment (Yancey1989) [reply vision/datasets/imageloader_test.go:96]: done

**Review by wangkuiyi — APPROVED** [review]

**Pull request merged**